tbtc-ts/src/sweep.ts

```diff
@@ -0,0 +1,107 @@
+import { BigNumber } from "ethers"
+import {
+  RawTransaction,
+  UnspentTransactionOutput,
+  Client as BitcoinClient,
+} from "./bitcoin"
+
+/**
+ * Sweeps UTXOs by combining all the provided UTXOs and broadcasting a Bitcoin
+ * P2WPKH sweep transaction.
+ * @dev The caller is responsible for ensuring the provided UTXOs are correctly
+ *      formed, can be spent by the wallet and their combined value is greater
+ *      then the fee.
+ * @param utxos - UTXOs to be combined into one output.
+ * @param fee - the value that should be subtracted from the sum of the UTXOs
+ *              values and used as the transaction fee.
+ * @param walletPrivateKey - Bitcoin private key of the wallet.
+ * @param bitcoinClient - Bitcoin client used to interact with the network.
+ * @returns Empty promise.
+ */
+export async function sweepUtxos(
+  utxos: UnspentTransactionOutput[],
+  fee: BigNumber,
+  walletPrivateKey: string,
+  bitcoinClient: BitcoinClient
+): Promise<void> {
+  const utxosWithRaw: (UnspentTransactionOutput & RawTransaction)[] = []
+  for (const utxo of utxos) {
+    const rawTransaction = await bitcoinClient.getRawTransaction(
+      utxo.transactionHash
+    )
+
+    utxosWithRaw.push({
+      ...utxo,
+      transactionHex: rawTransaction.transactionHex,
+    })
+  }
+
+  const transaction = await createSweepTransaction(
+    utxosWithRaw,
+    fee,
+    walletPrivateKey
+  )
+
+  await bitcoinClient.broadcast(transaction)
+}
+
+/**
+ * Creates a Bitcoin P2WPKH sweep transaction.
+ * @param utxos - UTXOs that should be used as transaction inputs.
+ * @param fee - the value that should be subtracted from the sum of the UTXOs
+ *              values and used as the transaction fee.
+ * @param walletPrivateKey - Bitcoin private key of the wallet.
+ * @returns Bitcoin sweep transaction in raw format.
+ */
+export async function createSweepTransaction(
+  utxos: (UnspentTransactionOutput & RawTransaction)[],
+  fee: BigNumber,
+  walletPrivateKey: string
+): Promise<RawTransaction> {
+  const decodedWalletPrivateKey = wif.decode(walletPrivateKey)
+
+  const walletKeyRing = new bcoin.KeyRing({
+    witness: true,
+    privateKey: decodedWalletPrivateKey.privateKey,
+    compressed: decodedWalletPrivateKey.compressed,
+  })
+
+  const walletAddress = walletKeyRing.getAddress("string")
+
+  const inputCoins = []
+  let totalInputValue = BigNumber.from(0)
+
+  for (const utxo of utxos) {
+    inputCoins.push(
+      bcoin.Coin.fromTX(
+        bcoin.MTX.fromRaw(utxo.transactionHex, "hex"),
+        utxo.outputIndex,
+        -1
+      )
+    )
+    totalInputValue = totalInputValue.add(utxo.value)
+  }
+
+  const transaction = new bcoin.MTX()
+
+  transaction.addOutput({
+    script: bcoin.Script.fromAddress(walletAddress),
+    value: totalInputValue.toNumber(),
+  })
+
+  await transaction.fund(inputCoins, {
+    changeAddress: walletAddress,
+    hardFee: fee.toNumber(),
+    subtractFee: true,
+  })
+
+  transaction.sign(walletKeyRing)
```

Comment:
That won't work. We have P2(W)SH outputs taken as inputs here. To unblock such inputs you must assemble an appropriate unlocking script for each input. 

For P2SH the locking script is `OP_HASH160 <redeemScript hash> OP_EQUAL` so the unlocking script must be `<something> <redeemScript>` where `something` is used during script computation and depends on the script content. The `redeemScript` is the serialized script used to lock the P2SH output which can be obtained using `createDepositScript` function.

For P2WSH, the locking script is `OP_0 <redeemScript hash>` and the unlocking script should look similar as in P2SH.